### PR TITLE
Implement integration test for gomod-vendor-check

### PR DESCRIPTION
**Pull request by dcho5**
CLOUDBLD-5578

Add test function which fails with inconsistent vendor or empty vendor.
Add test data for consistent vendor and no vendor.

Signed-off-by: Daniel Cho <dacho@redhat.com>

### CODE REVIEW

**Comment by dcho5** [comment]
I haven't implemented the test case for `empty_vendor` in `test_gomod_vendor_check_pass`, yet, but how does this look so far, @akhmanova ?

File: tests/integration/test_gomod_packages.py [hunk 1]
```diff
@@ -32,3 +32,73 @@ def test_gomod_vendor_without_flag(test_env):
...
+        "The content of the vendor directory is not consistent with go.mod. "
+        "Run `go mod vendor` locally to fix this problem. See the logs for more details."
+    )
+    assert error_msg in completed_response.data["state_reason"], (
+        f"#{completed_response.id}: Request failed correctly, but with unexpected message: "
+        f"{completed_response.data['state_reason']}. Expected message was: {error_msg}"
+    )
+
+
+def test_gomod_vendor_check_pass(test_env):
```
Comment (akhmanova) [inline tests/integration/test_gomod_packages.py:67]: @chmeliik 
> Would the empty_vendor case be the same as #501? Or is that something else?

Your example shows a gomod repo without deps. This question is mostly about a gomod repo with dependencies but with an empty vendor directory. I think it can be the same result as for `test_gomod_vendor_check_fai()`, because it is just a special case of the general. But I am not 100% sure and if @dcho5 has time, I would like it to be checked as well.

Comment (akhmanova) [reply tests/integration/test_gomod_packages.py:67]: WDYT? :)

File: tests/integration/test_gomod_packages.py [hunk 2]
```diff
@@ -32,3 +32,73 @@ def test_gomod_vendor_without_flag(test_env):
         )
     else:
         utils.assert_properly_completed_response(completed_response)
+
+
+def test_gomod_vendor_check_fail(test_env):
```
Comment (akhmanova) [inline tests/integration/test_gomod_packages.py:37]: The first test looks good to me! Thank you!

File: tests/integration/test_gomod_packages.py [hunk 3]
```diff
@@ -32,3 +32,73 @@ def test_gomod_vendor_without_flag(test_env):
...
+    )
+    assert error_msg in completed_response.data["state_reason"], (
+        f"#{completed_response.id}: Request failed correctly, but with unexpected message: "
+        f"{completed_response.data['state_reason']}. Expected message was: {error_msg}"
+    )
+
+
+def test_gomod_vendor_check_pass(test_env):
+    """
+    Validate successful completion of gomod vendor request with gomod-vendor-check flag.
```
Comment (akhmanova) [inline tests/integration/test_gomod_packages.py:69]: I think this test copies [the common one](https://github.com/release-engineering/cachito/blob/master/tests/integration/test_packages.py#L25). Could you please move the test data there or point me why you choose to have a new one?

Comment (dcho5) [reply tests/integration/test_gomod_packages.py:69]: I'll move the first case to `test_packages.py`, and keep the `empty_vendor` and `no_vendor` test cases in this file.

**Comment by dcho5** [comment]
Moved `correct_vendor` and `no_vendor` test cases to `test_packages` from `test_gomod_vendor_check_pass`. For, `no_vendor` an initial Cachito request is completed for `gomod-vendor-check-no-vendor`, which has dependencies in `go.mod` but no vendor directory. The test checks that the request is successful, and that the cached files match those of `gomod-vendor-check-pass` (which has a consistent vendor directory) instead of `gomod-vendor-check-no-vendor`.

File: tests/integration/test_gomod_packages.py (hunk 3 above)
Comment (akhmanova) [reply tests/integration/test_gomod_packages.py]: Great, thank you

**Comment by dcho5** [comment]
Implemented all tests, please review @akhmanova

File: tests/integration/test_gomod_packages.py (hunk 1 above)
Comment (chmeliik) [reply tests/integration/test_gomod_packages.py]: I think it's the same case as `test_gomod_vendor_check_fail`, yeah. But doesn't hurt to have both :-)

File: tests/integration/test_data/gomod_vendor_check.yaml [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+# expected_deps_files: Expected dependencies files (empty)
+# response_expectations: Parts of the Cachito response to check
+# content_manifest: PURLs for image contents part
+wrong_vendor:
+  repo: https://github.com/cachito-testing/gomod-vendor-check-fail.git
+  ref:  8553df6498705b2b36614320ca0c65bc24a1d9e6
+  flags: ["gomod-vendor-check"]
+  pkg_managers: ["gomod"]
+empty_vendor:
+  repo: https://github.com/cachito-testing/gomod-vendor-check-empty-vendor.git
```
Comment (chmeliik) [inline tests/integration/test_data/gomod_vendor_check.yaml:15]: I expected that https://github.com/cachito-testing/gomod-vendor-check-empty-vendor would have an empty `vendor/` directory. Would that be a more interesting test case than just having empty `modules.txt`? IMO that is very similar to https://github.com/cachito-testing/gomod-vendor-check-fail

Note: it can't be completely empty, since git doesn't recognize empty dirs. You would have to put something like an empty `.gitkeep` file in there

File: tests/integration/test_gomod_packages.py [hunk 5]
```diff
@@ -32,3 +34,34 @@ def test_gomod_vendor_without_flag(test_env):
...
+            "flags": env_data["flags"],
+            "pkg_managers": env_data["pkg_managers"],
+        },
+    )
+    completed_response = client.wait_for_complete_request(initial_response)
+    assert completed_response.status == 200
+    assert completed_response.data["state"] == "failed"
+    error_msg = (
+        "The content of the vendor directory is not consistent with go.mod. "
+        "Run `go mod vendor` locally to fix this problem. See the logs for more details."
```
Comment (chmeliik) [inline tests/integration/test_gomod_packages.py:62]: It might also be interesting to check the logs here, since we promise there will be more details :D

AFAIK it's not something we've done in integration tests yet so I'd be OK with leaving it as a future improvement, I'll leave it up to you.

**Review by chmeliik — COMMENTED** [review]
Nice! I've got some suggestions :-)

File: tests/integration/test_data/gomod_vendor_check.yaml (hunk 4 above)
Comment (dcho5) [reply tests/integration/test_data/gomod_vendor_check.yaml:15]: That's exactly what I thought initially as well, but I also knew that git doesn't recognize empty dirs. I wasn't aware of the `.gitkeep` trick. I'll give it a try!

Comment (dcho5) [reply tests/integration/test_data/gomod_vendor_check.yaml:15]: Done!

File: tests/integration/test_gomod_packages.py (hunk 5 above)
Comment (dcho5) [reply tests/integration/test_gomod_packages.py:62]: Were you thinking that we confirm the details of the log (the same way we confirm the `error_msg`) in the test function as well? The logs we are looking for are logged in `_vendor_changed` - how can I access the logs in `test_gomod_vendor_check_fail`?

Comment (chmeliik) [reply tests/integration/test_gomod_packages.py:62]: Yeah that's what I was thinking. Request logs can be accessed at a url, e.g. localhost:8080/api/v1/requests/1/logs

File: tests/integration/test_data/gomod_vendor_check.yaml (hunk 4 above)
Comment (chmeliik) [reply tests/integration/test_data/gomod_vendor_check.yaml:15]: Nice :+1:

**Review by chmeliik — APPROVED** [review]
LGTM :tada: 

Checks for logs can be added later as we discussed

**Review by akhmanova — APPROVED** [review]
